File: src/azisaba/semaphore/spigot/SpigotSemaphore.scala
package azisaba.semaphore.spigot

import azisaba.semaphore.spigot.hook.QuitEventDataSaveHook
import azisaba.semaphore.spigot.listener.PlayerQuitListener
import org.bukkit.event.HandlerList
import org.bukkit.plugin.java.JavaPlugin

import scala.collection.mutable

/*
 * @author amata1219
 */

class SpigotSemaphore extends JavaPlugin() {

  SpigotSemaphore.instance = this

  override def onEnable(): Unit = {
    getServer.getPluginManager.registerEvents(SpigotSemaphore.hookRegistry, this)
  }

  override def onDisable(): Unit = {
    HandlerList.unregisterAll(this)
  }

}
object SpigotSemaphore {

  var instance: SpigotSemaphore = _

  val hookList: mutable.ArrayBuffer[QuitEventDataSaveHook[_]] = mutable.ArrayBuffer()

  // フックの登録先として機能するイベントリスナ
  val hookRegistry = new PlayerQuitListener()

}

File: src/main/scala/azisaba/semaphore/spigot/SpigotSemaphore.java
package azisaba.semaphore.spigot;

import amata1219.redis.plugin.messages.spigot.RedisPluginMessagesAPI;
import azisaba.semaphore.spigot.hook.SynchronousDataSaveHook;
import azisaba.semaphore.spigot.listener.PlayerQuitListener;
import azisaba.semaphore.spigot.redis.RedisPluginMessaging;
import azisaba.semaphore.spigot.redis.RedisPublisher;
import org.bukkit.event.HandlerList;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.java.JavaPlugin;

public class SpigotSemaphore extends JavaPlugin {

    public static final String PLUGIN_MESSAGING_CHANNEL = "BungeeSemaphore";

    private static SpigotSemaphore instance;

    private final PlayerQuitListener hookRegistry = new PlayerQuitListener();
    private RedisPublisher publisher;

    @Override
    public void onEnable() {
        instance = this;

        Plugin maybeRedisPluginMessages = getServer().getPluginManager().getPlugin("RedisPluginMessages");
        if (!(maybeRedisPluginMessages instanceof RedisPluginMessagesAPI))
            throw new NullPointerException("Not found RedisPluginMessages plugin.");

        publisher = new RedisPluginMessaging((RedisPluginMessagesAPI) maybeRedisPluginMessages);

        getServer().getPluginManager().registerEvents(hookRegistry, this);
    }

    @Override
    public void onDisable() {
        HandlerList.unregisterAll(this);
    }

    public static SpigotSemaphore isntance() {
        return instance;
    }

    public RedisPublisher publisher() {
        return publisher;
    }

    public void registerHook(JavaPlugin hostPlugin) {
        SynchronousDataSaveHook.registerHook(hookRegistry, hostPlugin);
    }

}
